networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BlobSidecarsByRangeMessageHandler.java

```diff
@@ -127,12 +132,31 @@ public void onIncomingMessage(
               if (checkRequestInMinEpochsRange(requestEpoch)
                   && !checkBlobSidecarsAreAvailable(earliestAvailableSlot, endSlot)) {
                 return SafeFuture.failedFuture(
-                    new RpcException.ResourceUnavailableException(
-                        "Requested blob sidecars are not available."));
+                    new ResourceUnavailableException("Requested blob sidecars are not available."));
               }
-              final BlobSidecarsByRangeMessageHandler.RequestState initialState =
-                  new BlobSidecarsByRangeMessageHandler.RequestState(
-                      callback, maxRequestBlobSidecars, startSlot, endSlot);
+
+              final UInt64 finalizedSlot =
+                  combinedChainDataClient.getFinalizedBlockSlot().orElse(UInt64.ZERO);
+
+              final SortedMap<UInt64, Bytes32> canonicalHotRoots;
+              if (endSlot.isGreaterThan(finalizedSlot)) {
+                final UInt64 count = endSlot.minus(finalizedSlot);
+                final UInt64 hotBlocksStartSlot = finalizedSlot.plus(1);
+
+                canonicalHotRoots =
```

Comment:
I'm a little worried on concurrency here, but looking at code, if we have new finalized epoch between this line and finalizedSlot above, we just get empty hotRoots and return smaller set to the requester, maybe even empty. Not a fatal, just noticing. Though we should reply with at least one element as I remember by spec. Not a big bug I think anyway

Reply:
Good point!
I should have improved that.

Reply:
Sorry I saw your comment after selecting auto merge. What i'm more concerned about is to get a consistent finalizedSlot-canonicalRoots pair instead of being concerned about get them updated.
In theory the definitive solution is to have a single method that atomically get hotRoots and the finalized slot.
If things change while we iterate it can always happen but i actually accepted that for performance reason (it was actually my first draft and i moved away from it)

Reply:
No problem! I totally like it as a first iteration. Let's see how it goes